[libpas] Add alloc/dealloc stack traces to WebKit PGM

zakariaridouh opened the pull request:
#### 1dc8717af8a79287e871a499e6ca3cbd32fde51b
<pre>
[libpas] Add alloc/dealloc stack traces to WebKit PGM
<a href="https://bugs.webkit.org/show_bug.cgi?id=283370">https://bugs.webkit.org/show_bug.cgi?id=283370</a>
<a href="https://rdar.apple.com/106166552">rdar://106166552</a>

Reviewed by David Kilzer.

This adds collection and storage of &lt;=31 frames of backtrace upon PGM
allocation and deallocation. A followup patch will pack these frame
pointers more aggressivly.

Additionally, bump `pas_crash_report_version` to version 3 so
`reportcrashd` is aware of struct changes.

Add a variety of tests to validate this new functionality.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMBmallocAllocationBacktrace):
(std::testPGMAllocMetadataOnly):
(addPGMTests):

Canonical link: <a href="https://commits.webkit.org/289218@main">https://commits.webkit.org/289218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f3645d8108220986559fd78417056fdb7cd75f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85900 "34 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5520 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40268 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/90909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/36813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13524 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/90909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/36813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88903 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4395 "Build is in progress. Recent messages:") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77904 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/90909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/4262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32186 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/35892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/78838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33045 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/84822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13150 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/9646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/92632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73761 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18377 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/17220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13181 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107243 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/12958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/25827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/16391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/14745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
```diff
@@ -51,18 +51,32 @@
 
 PAS_BEGIN_EXTERN_C;
 
+#ifndef MAX_FRAMES
```
Why are we doing `#ifndef` here?

zakariaridouh replied:
Removed.

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
I would like a bit more of a descriptive name.

stwrt replied:
Something like `PGM_BACKTRACE_MAX_FRAMES`?

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -28,6 +28,7 @@
 
 #if LIBPAS_ENABLED
 
+#include <execinfo.h>
```
Playstation doesn't seem to have this header.

```
pas_probabilistic_guard_malloc_allocator.c:31:10: fatal error: 'execinfo.h' file not found
```

zakariaridouh replied:
After discussion with the PlayStation folks, since they don't have the `backtrace` API implemented, we should just no-op on PlayStation. We can revisit afterwards if they decide to implement this API.

rkirsling replied:
Doesn't seem like this was done; you just broke us instead. 😐

zakariaridouh replied:
fixing here: https://bugs.webkit.org/show_bug.cgi?id=286341

Sorry about that! If you look at `pas_probabilistic_guard_malloc_allocator.h:64` I gated the PlayStation build from importing `<execinfo.h>`. Seems like that didn't work.

Pushing a fix now.

cc: @rkirsling

rkirsling replied:
Ah, I reverted it in https://github.com/WebKit/WebKit/commit/725b72bb34b3437e1c1503fb8301608961cedbb6 to get our EWS unstuck.

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -43,6 +43,13 @@
+#include <execinfo.h>
+#else
+size_t backtrace(void **, size_t) { return 0; }
+#endif
```
This is fine as-is, but a slightly better way to do this (in the future) would be to define a `PAS_HAVE(BACKTRACE)` or `PAS_HAVE(EXECINFO)` macro, and define that based on the platform.

No requirement to change this before landing, but having a more descriptive macro means you can remove the comment here and again in `PGMTests.cpp`.  :)

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
```diff
@@ -55,18 +55,30 @@
 #include "pas_utils.h"
 #include "pas_large_heap.h"
 #include "pas_ptr_hash_map.h"
+#include "pas_large_map_entry.h"
```
Nit: Please fix alphabetical order of includes—at least for these last three items.

ddkilzer commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -20,14 +20,17 @@
  */
 
-#include <unistd.h>
+#include <execinfo.h>
```
Technically, you should exclude this for `#if !PAS_PLATFORM(PLAYSTATION)`/`#endif`.

I'm not sure if it matters in practice.  (Does that platform builds this test?)

zakariaridouh replied:
I don't believe so, but good catch. Excluded.

ddkilzer reviewed (approved):
r=me

Only need to fix the header ordering issue and the missing `#if` around `#include <execinfo.h>`.

The other change is optional, but would remove a couple of comments with self-documenting macro definitions.

stwrt commented on Source/bmalloc/bmalloc/BPlatform.h:
```diff
@@ -342,11 +342,11 @@
 #ifndef PAS_BMALLOC
 #define PAS_BMALLOC 1
-#endif
+#endif // BENABLE_LIBPAS 1
```
These updates aren't really part of the main PR.

zakariaridouh committed:
[libpas] Add alloc/dealloc stack traces to WebKit PGM
https://bugs.webkit.org/show_bug.cgi?id=283370
rdar://106166552

Reviewed by David Kilzer.

This adds collection and storage of <=31 frames of backtrace upon PGM
allocation and deallocation. A followup patch will pack these frame
pointers more aggressivly.

Additionally, bump `pas_crash_report_version` to version 3 so
`reportcrashd` is aware of struct changes.

Add a variety of tests to validate this new functionality.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMBmallocAllocationBacktrace):
(std::testPGMAllocMetadataOnly):
(addPGMTests):

Canonical link: https://commits.webkit.org/289218@main
